### ISSUE TITLE
accounts_db indexes taking too long be created

### ISSUE THREAD

**Issue by joeaba** [solana-labs/solana#17019]
#### Problem
Nodes with the `--account-index` argument will take too much to create the accounts_db indexes.
n2-standard-32 (32 vCPUs, 128 GB memory) servers will take a ~ 4 hours.
n2-standard-64 (64 vCPUs, 256 GB memory) servers will take from 30 mins to 1 hour.

Here are the logs from a fresh n2-standard-64  setup:

[solana-validator.log](https://github.com/solana-labs/solana/files/6418830/solana-validator.log)
[LOG.txt](https://github.com/solana-labs/solana/files/6418831/LOG.txt)

It's been up for more than 2 hours now and it's still at `100037/381300`

Let me know if you need more logs.

**Comment by t-nelson** [comment]
Interesting.  Are the `n2-std-32`s swapping a bunch or something?

**Comment by joeaba** [comment]
Yes, it's sucking up all the ram

**Comment by mvines** [comment]
cc: @sakridge @jeffwashington @carllin

**Comment by carllin** [comment]
I think we can probably save a lot of memory by:

1. Removing the `reverse_index` here: https://github.com/solana-labs/solana/blob/0ebfa08860bc3003e7d2973df03abaf34e83d188/runtime/src/secondary_index.rs#L148
2.  Removing the `HashSet<Slot>` here: https://github.com/solana-labs/solana/blob/0ebfa08860bc3003e7d2973df03abaf34e83d188/runtime/src/secondary_index.rs#L25

These were really only used to figure out in `purge_secondary_indexes_by_inner_key()` when the *inner* account pubkeys from the index could be dropped. For instance, if we had the secondary index `Mint pubkey -> Account pubkeys`, it would figure out when those inner account pubkeys could be dropped. 

However, it seems like the call to `purge_secondary_indexes_by_inner_key()` from `handle_dead_keys()` here: should be sufficient https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_index.rs#L833-L837.

Currently reviewing my old PR here to figure out if I'm missing something/why this extra complexity was needed. 

@ryoqun made these comments here: https://github.com/solana-labs/solana/pull/14212#discussion_r546342055 and here: https://github.com/solana-labs/solana/pull/14212#discussion_r546340527 which speaks to this. Perhaps we were thinking we could filter unrooted slots when scanning the secondary index? I don't think that functionality is currently used here in the scan: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_index.rs#L759, we grab all the pubkeys, regardless of the slot in the secondary index.

**Comment by carllin** [comment]
Other suggestions:

1) Maybe only certain nodes should have certain certain `SecondaryIndexes`?

2) We only have the `SecondaryIndex` for large keysets like the token mint for specific keys? I.e. partners will have to run their own nodes if they want an index for something large like their Tokenmint

3) Discuss we can offload portions of the AccountsIndex and SecondaryIndexes to disk. Essentially I think this comes down to how to efficiently update and represent an equivalent to `HashMap<Pubkey, HashSet<Pubkey>>` on disk. Key-value store like Rocks with a key == `(Pubkey, Pubkey)` and a prefix iterator perhaps 😅

**Comment by carllin** [comment]
Update: Some of these SecondaryIndexes seems a lot bigger than I would've expected, i.e. the `spl_token_mint_index` seems to be about 40 GB currently, about 4-5x bigger than expected. Might be the overhead of some of these `DashMap`'s...

**Comment by joeaba** [comment]
Removing `spl_token_mint_index` fixes the problem.

**Comment by sakridge** [comment]
> Removing `spl_token_mint_index` pretty much fixes the problem.

Thanks for narrowing this down.

**Comment by jeffwashington** [comment]
I have been able to reproduce this. However, I'm confused with the action items from the discussion and options above. @joeaba says removing spl_token_mint_index allows validators to start, but that some method calls will now be very slow. For fresh setups, this will not work.

**Comment by carllin** [comment]
@jeffwashington I think ultimately we should try experimenting with offloading these secondary indexes (essentially `SecondaryIndex structure, which is essentially a `HashMap<Pubkey, HashSet<Pubkey>>`) to disk. This generally means something like a disk-based Btree or LSM tree.


An experiment I think that is worth trying is moving these SecondaryIndexes to RocksDb, where the key for an account in the mint index would look something like`(Mint Pubkey, Account Pubkey)` and we can run prefix searches to find all the accounts for a given mint.

Pros I can think of:

1) Offloads unused keys to disk
2) Heavy write path is optimized since all deletions/additions happen in memory, so it's fast, and immediately observable.
3) Memtables and SST files allow concurrent read/write b/c they are based on a concurrent skiplist (a robust implementation of which is currently unavailable in Rust), so we should be able to support RPC scan through the secondary index without blocking writes from replay, where DashMap currently blocks large chunks of accounts from updating while scanning over the shards in the DashMap here: https://github.com/solana-labs/solana/blob/master/runtime/src/secondary_index.rs#L66-L68.

**Comment by sakridge** [comment]
@carllin if not using rocksdb, a simple way could be to keep the HashMap<Pubkey, File> in memory. Then store each HashSet as a file. Could be just a list of pubkeys in a file, maybe with a byte to invalidate one if needed, then it may have to compact if necessary.

**Comment by carllin** [comment]
> @carllin if not using rocksdb, a simple way could be to keep the HashMap<Pubkey, File> in memory. Then store each HashSet as a file. Could be just a list of pubkeys in a file, maybe with a byte to invalidate one if needed, then it may have to compact if necessary.

Yeah, this was our initial thought as well, but some of the downsides I could see:

1) With large `HashSets` like the program id index entry for Kin for instance, lookups for specific keys for deletion/existence check will incur a large scan of the file
2) Writes to the individual hashsets on disk can be a bit slower
3) Maintaining consistency for scans while reading the files and compacting could be tricky to implement. Without some of the lock-free utilities from C++ in Rust, might also have to block either the scan or the compaction, so our implementation would likely cause more contention than the one in Rocks.

Might still be worth pursuing and benchmarking though.

**Comment by jeffwashington** [comment]
I can work on the simple way and get some benchmarks. That would get me some experience here and us some data.

**Comment by CriesofCarrots** [comment]
Yeah, I think the RocksDb experiment is worth benchmarking. The concurrent read/write support seems particularly worthwhile. Seems like it should perform pretty well once the node is running, but I'm a little concerned that checking/updating the on-disk index on boot won't be appreciably faster than building the in-memory index.

**Comment by jeffwashington** [comment]
I will begin this prototype wed. morning. Talked with @carllin , @steviez , and @brooksprumo

**Comment by sakridge** [comment]
Some other "easy" solutions:
1. Async index generation - have an argument to accounts_index::insert_new_entry_if_missing to not add the secondary indexes on `AccountsDB::generate_index`. Then start a thread to scan the accounts set to add them after the validator is starting up. RPC access with scans will be slow until the index is fully populated and usable, but the validator wont' be so far behind. The secondary indexes would have to have a state where they are not yet usable and ready.

2. Start up the network stack right after snapshot download. Problem is a lot of the services want a bank to work more efficiently. bank_forks would have to be substituted in later. Idea is the validator at least is downloading shreds while it's also dealing with the snapshot/replay work.

For the on-disk index solutions 1. might be required anyway depending on how long that takes to generate.

**Comment by jeffwashington** [comment]
#17040 helps
#17048 would help

**Comment by jeffwashington** [comment]
I talked with @sakridge . I will first do async index generation. Then, rocksdb secondary index. Steven will investigate the customer use cases.

**Comment by mvines** [comment]
(Longer term "Start up the network stack right after snapshot download" would be amazing.  Even before snapshot download, we can opportunistically insert shreds that we receive from the network.  Those opportunistic shreds would still need to be validated against the leader schedule once the snapshot is loaded.   This is a medium-sided project and out of scope of this particular issue though!)

**Comment by sakridge** [comment]
This is the expensive key I think:
```
[2021-05-05T18:23:19.450413579Z INFO  solana_runtime::accounts_index] spl_token_mint_index: kinXdEcpDQeHPEuQnqmUgtYykqKGVFq6CeVX5iAHJq6 (
        60789809,
        60818121,
    )
```

**Comment by jeffwashington** [comment]
Does kin need to call and use the secondary index? Could we have mint key(s) we pass as args to validator to say: skip indexing this index? Kind of hacky, but low risk in the short term, configurable, requires barely any code change, and fits our real world use case for now, I think. Alternatively, we could specify the opposite - the mint keys we DO care to index on THIS rpc machine.

**Comment by jeffwashington** [comment]
Another aspect of a solution is figuring out we will fail a query/scan due to results being too large and erroring out.

**Comment by jeffwashington** [comment]
Another aspect is reducing contention between validator's normal operation and the large scans so the validator doesn't fall behind.

**Comment by sakridge** [comment]
Another idea is doing just a basic split of the snapshot files. This has the advantage that any one download failure is not as expensive. The validator could also then pipeline the download and snapshot ingestion work.

**Comment by jeffwashington** [comment]
async secondary indexes: #17090
